docs/adrs/008-unified-inputs.md

```diff
@@ -0,0 +1,106 @@
+Input := [FromSource InputPath, FromJob { job : Job, files : List InputPath }]
+
+# Add the given file to the job's workspace (the working directory where the
+# command is called.)
+file : Str -> Input
+file = \inputPath -> FromSource (path inputPath)
+
+# Set up a file from the project source (first argument) in exactly the location
+# you need it within the workspace (second argument.)
+renamedFile : Str, Str -> Input
+renamedFile = \from, to -> FromSource (rename from to)
+
+# Put all of the outputs of the given job in the current job's workspace.
+#
+# This is the easiest way to source files from other jobs. That said, if you have
+# multiple jobs with the same file this can cause conflicts. In that case, use
+# `someOutputsOf` to tell me exactly what you want.
+allOutputsOf : Job -> Input
+allOutputsOf = \Job config -> FromJob { job: Job config, files: List.map path job.outputs }
+
+# Use this function to set up a the files from another job exactly the way you
+# want in your workspace. You can pick exatly the files you need like
+# `someOutputsOf someJob [file "the-one-I-want"]` or put them exactly in the
+# locations you need like `someOutputsOf someJob [rename
+# "path-that-doesnt-work-for-me" "a/better/path"]`.
+someOutputsOf : Job, List InputPath -> Input
+someOutputsOf = \job, files -> FromJob { job, files }
+
+InputPath := [Path Str, RenamePath { from : Str, to : Str }]
+
+# Use the path exactly as given, with no shenanigans.
+path : Str -> InputPath
+path = \str -> Path str
+
+# Make the specified input path (first argument) available at a different
+# location in the workspace (second argument.)
+rename : InputPath, Str -> InputPath
+rename = \from, to -> ...
+```
+
+Usage might look like this:
+
+```coffeescript
+completedGreeting : Job
+completedGreeting =
+    job {
+        command: exec (systemTool "bash") [
+            "-c",
+            "printf '%s, %s!\n' "$(cat greeting.txt)" "$(cat subject.txt)" > completedGreeting.txt",
+        ],
+        inputs: [
+            allOutputsOf subject,
+            someOutputsOf greeting [rename (file "englishGreeting.txt") "greeting.txt"],
+        ],
+        outputs: ["completedGreeting.txt"]
+    }
+
+subject : Job # has `subject.txt` in its outputs
+
+greeting : Job # has at least `englishGreeting.txt` in its outputs
+```
+
+This can be extended with filesystem matchers (our version of glob matches) and dynamic dependency discoverers later, without breaking the API.
+(As a matter of fact, matchers might take over `FromSource`'s job eventually!)
+
+## Benefits
+
+This API:
+
+- Lets us specify exactly how jobs want the filesystem to be set up before they run
+- Lets us warn about conflicts between different files.
+- Lets us see exactly where we're trying to source files from.
+  This means we can see when a file would not exist and warn about that.
+  (`allOutputsOf` defeats this a bit since it automatically depends on whatever files are available, but overall I think it's a win.)
```

Comment:
Based on most of the well setup build systems I have used, `allOutputsOf` is essentially never the correct option. It is just used because it is convenient. I am not sure I would call it a win, but I can see the want for it.

Reply:
yeah… I'm not 100% certain of it either. I think it's probably possible to do something nicer once we have globs/file matchers. At that point, if you really want everything you can specify it pretty easily, but you at least need to think about what you'd like.